File: objectiveai-js/src/chat/completions/response/unary/chat_completion.ts
import z from "zod";
import { ChoiceSchema } from "./choice";
import { UpstreamSchema } from "../../upstream";
import { UsageSchema } from "../usage";
import { ResponseObjectSchema } from "./response_object";
import { convert, type JSONSchema } from "../../../../json_schema";

export const ChatCompletionSchema = z
  .object({
    id: z.string().describe("The unique identifier of the chat completion."),
    upstream_id: z
      .string()
      .describe("The unique identifier of the upstream chat completion."),
    choices: z
      .array(ChoiceSchema)
      .describe("The list of choices in this chat completion."),
    created: z
      .uint32()
      .describe(
        "The Unix timestamp (in seconds) when the chat completion was created."
      ),
    model: z
      .string()
      .describe(
        "The unique identifier of the Ensemble LLM used for this chat completion."
      ),
    upstream_model: z
      .string()
      .describe("The upstream model used for this chat completion."),
    object: ResponseObjectSchema,
    service_tier: z.string().optional(),
    system_fingerprint: z.string().optional(),
    usage: UsageSchema,
    upstream: UpstreamSchema,
    provider: z
      .string()
      .optional()
      .describe("The provider used for this chat completion."),
  })
  .describe("A unary chat completion response.");
export type ChatCompletion = z.infer<typeof ChatCompletionSchema>;
export const ChatCompletionJsonSchema: JSONSchema = convert(ChatCompletionSchema);

File: objectiveai-rs/src/chat/completions/response/unary/chat_completion.rs
//! Chat completion response type.

use crate::chat::completions::response;
use serde::{Deserialize, Serialize};

/// A complete chat completion response.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ChatCompletion {
    /// ObjectiveAI's unique identifier for this completion.
    pub id: String,
    /// The upstream provider's identifier.
    pub upstream_id: String,
    /// The generated choices.
    pub choices: Vec<super::Choice>,
    /// Unix timestamp when the completion was created.
    pub created: u64,
    /// The Ensemble LLM ID used.
    pub model: String,
    /// The upstream model identifier.
    pub upstream_model: String,
    /// The object type (always "chat.completion").
    pub object: super::Object,
    /// The service tier used, if applicable.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub service_tier: Option<String>,
    /// A fingerprint of the model configuration.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub system_fingerprint: Option<String>,
    /// Token usage statistics.
    pub usage: response::Usage,
    /// Upstream provider
    pub upstream: crate::chat::completions::Upstream,

    /// The provider that served the request (OpenRouter-specific).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
}

impl From<response::streaming::ChatCompletionChunk> for ChatCompletion {
    fn from(
        response::streaming::ChatCompletionChunk {
            id,
            upstream_id,
            choices,
            created,
            model,
            upstream_model,
            object,
            service_tier,
            system_fingerprint,
            usage,
            upstream,
            provider,
        }: response::streaming::ChatCompletionChunk,
    ) -> Self {
        Self {
            id,
            upstream_id,
            choices: choices.into_iter().map(super::Choice::from).collect(),
            created,
            model,
            upstream_model,
            object: object.into(),
            service_tier,
            system_fingerprint,
            usage: usage.unwrap_or_default(),
            upstream,
            provider,
        }
    }
}
